### PR TITLE
adding undo/redo

### PR DESCRIPTION
**Pull request by rodydavis**
- Undo Last Change
- Redo Previous Undo
- Clear Undo/Redo History

### CODE REVIEW

**Comment by mono0926** [comment]
It adds additional cost(mainly memory).
How about implement as a mixin?

**Comment by rodydavis** [comment]
It can have a Mac set on the tracked changes. And not sure if it can be a mixin as I need to intercept the setter on value

**Comment by rodydavis** [comment]
I added 0 to the tracked changes by default. This means that it will have not memory impact. By adding -1 to `maxTrackedChanges` it will have an unbounded history and a number greater than 0 will remove the oldest change from the stack when it reaches the max.

File: packages/state_notifier/lib/state_notifier.dart [hunk 1]
```diff
@@ -73,9 +75,22 @@ typedef Locator = T Function<T>();
...
+  /// If [-1] then it will have an unbounded history.
+  ///
+  /// If [0] then it will not track changes.
+  /// This is the default option.
+  ///
+  /// If [>=1] then it will throw away the oldest change.
+  int maxTrackedChanges;
 
   final _listeners = LinkedList<_ListenerEntry<T>>();
+  ChangeStack<T> _changeStack;
```
Comment (rrousselGit) [inline packages/state_notifier/lib/state_notifier.dart:93]: I'd go for a LinkedList, with the LinkedListEntry being the current position in the queue.
Undo goes to the previous node, redo to the next one
Then, when a new update is emitted, it's added at the end of the list, and that new entry is now the current position

**Comment by rrousselGit** [comment]
We'll want some tests for that + documentation
Also, if not for a mix-in, we can have a different base class. Like `UndoableStateNotifier`.

**Comment by rodydavis** [comment]
What about `TrackedStateNotifier`? I can create a new base class and write some tests

**Pull request closed without merging**